### PR TITLE
APP-2998 Make OBO configuration optional

### PR DESCRIPTION
**Pull request by symphony-elias**
Don't know how to test SymphonyBdk as it has many dependencies.

### CODE REVIEW

**Comment by symphony-youri** [comment]
> Don't know how to test SymphonyBdk as it has many dependencies.

Is the problematic dependency the ApiClientFactory? In that case I would create a specific constructor (not public) that takes it as a parameter to inject a mock

**Comment by symphony-elias** [comment]
> > Don't know how to test SymphonyBdk as it has many dependencies.
> 
> Is the problematic dependency the ApiClientFactory? In that case I would create a specific constructor (not public) that takes it as a parameter to inject a mock

Indeed but it means we will have to either duplicate code in the constructor or remove the `final` keywords on the fields.

**Comment by symphony-youri** [comment]
Like that it should be ok (we won't be able to get 100% line coverage though :))
```
public SymphonyBdk(BdkConfig config) throws AuthInitializationException, AuthUnauthorizedException {
    this(config, new ApiClientFactory(config));
  }

  SymphonyBdk(BdkConfig config, ApiClientFactory apiClientFactory)
      throws AuthInitializationException, AuthUnauthorizedException {

    this.apiClientFactory = apiClientFactory;

    final AuthenticatorFactory authenticatorFactory = new AuthenticatorFactory(config, this.apiClientFactory);

    this.botSession = authenticatorFactory.getBotAuthenticator().authenticateBot();
    this.oboAuthenticator = authenticatorFactory.getOboAuthenticator();

    // setup the datafeed
    final DatafeedApi datafeedApi = new DatafeedApi(this.apiClientFactory.getAgentClient());
    if (DatafeedVersion.of(config.getDatafeed().getVersion()) == DatafeedVersion.V2) {
      this.datafeedService = new DatafeedServiceV2(datafeedApi, this.botSession, config);
    } else {
      this.datafeedService = new DatafeedServiceV1(datafeedApi, this.botSession, config);
    }
  }
```

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/SymphonyBdk.java [hunk 1]
```diff
@@ -74,4 +74,11 @@ public MessageService messages(Obo.Handle oboHandle) throws AuthUnauthorizedExce
   public DatafeedService datafeed() {
     return this.datafeedService;
   }
+
+  protected OboAuthenticator getOboAuthenticator() throws AuthInitializationException {
+    if (this.oboAuthenticator == null) {
+      throw new AuthInitializationException("OBO is not configured", null);
+    }
+    return this.oboAuthenticator;
+  }
```
Comment (symphony-thibault) [inline symphony-bdk-core/src/main/java/com/symphony/bdk/core/SymphonyBdk.java:83]: I'd write something more Java8 :) 
```java
  protected OboAuthenticator getOboAuthenticator() {
    return Optional.ofNullable(this.oboAuthenticator)
        .orElseThrow(() -> new IllegalStateException("OBO is not configured."));
  }
```

As well, the `AuthInitializationException` is not really designed to handle this kind of error. Either we create a dedicated exception, or we keep the `IllegalStateException`. Anyway, that's the kind of error that a developer should see right when starting the application

Comment (symphony-youri) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/SymphonyBdk.java:83]: At the cost of allocating memory for the optional! :p

Comment (symphony-thibault) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/SymphonyBdk.java:83]: Aren't you ex-C++ dev?!

**Review by symphony-thibault — APPROVED** [review]
LGTM 👍

**Pull request merged**